pkg/injector/patcher/patch.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2021 Layotto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package patcher
+
+import (
+	"encoding/json"
+
+	jsonpatch "github.com/evanphx/json-patch/v5"
+)
+
+const (
+	// Path for patching containers.
+	PatchPathContainers = "/spec/containers"
```

Comment:
`/spec/initContainers` ?

Reply:
/cc @wenxuwan Should we use layotto as an init-container?

Reply:
I checked the logs of these two containers. It may be because when the client demo sent the request, the Layotto container had not yet started the GRPC service. The following are the  of the two containers:

1. client demo container logs
```
runtime client initializing for: 127.0.0.1:34904
Start testing state_demo
panic: error saving state: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:34904: connect: connection refused"
goroutine 1 [running]:
main.testSave({0x570dd8?, 0x40000a2000?}, {0x574d98?, 0x40000d4500?}, {0xffffc01fbbba?, 0x1?}, {0x4bc5f7, 0x4}, {0x40000a2dd0, 0xb, ...})
        /app/client.go:138 +0x110
main.main()
        /app/client.go:63 +0x1c8
```

2. Layotto container logs
```
...
2023-10-19 12:26:03,669 [INFO] [runtime] init PhoneCallService
2023-10-19 12:26:03,669 [INFO] [runtime] init SmsService
2023-10-19 12:26:03,669 [INFO] [runtime] initRuntime stages cost: 50.84825ms
2023-10-19 12:26:03,671 [ERROR] [runtime][listTopicSubscriptions]error after callback: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.65.254:9999: connect: connection refused"
2023-10-19 12:26:03,673 [DEBUG] grpc server filter initialized success
2023-10-19 12:26:03,673 [DEBUG] [config] AddOrUpdateNetworkFilterFactories store network filter factories, name: grpc
2023-10-19 12:26:03,674 [INFO] [server] [conn handler] [add listener] add listener: 0.0.0.0:34904
```

Reply:
I tested locally and use layotto as an init-container doesn't seem to work. The main container needs to wait for init-container to exit before it can be created. As a result, the main container is always in a pending state.

```
NAME                                      READY   STATUS      RESTARTS        AGE
layotto-state-sdk-demo-64b7c95966-49ktz   0/1     Init:0/1    0               8s
layotto-state-sdk-demo-64b7c95966-lm8l2   0/1     Init:0/1    0               8s
layotto-state-sdk-demo-64b7c95966-tb8kc   0/1     Init:0/1    0               8s
```

Reply:
You are correct, so can you update your demo, I checked that you directly panic when creating client failed, can we retry it for some time, there are some package can do such things.

Reply:
@Xunzhuo I checked the source code of layotto `client.NewClient()` and found that it internally calls `grpc.Dial` to create a connection. No GRPC call is sent.
```
conn, err := grpc.Dial(address, grpc.WithInsecure())
if err != nil {
	return nil, errors.Wrapf(err, "error creating connection to '%s': %v", address, err)
}
```
According to the client demo container logs, the panic is located in `cli.SaveState(ctx, store, key, value)`. So I used [retry-go](https://github.com/avast/retry-go) to retry after the request fails.
```
func testSave(ctx context.Context, cli client.Client, store string, key string, value []byte) {
	err := retry.Do(func() error {
		return cli.SaveState(ctx, store, key, value)
	},
		retry.Attempts(3),
		retry.Delay(time.Second),
	)
	if err != nil {
		panic(err)
	}
	fmt.Printf("SaveState succeeded.key:%v , value: %v \n", key, string(value))
}
```
Do you think this is reasonable?

Reply:
Tried and it worked well, thanks.